### ISSUE TITLE
About linear regression

### ISSUE THREAD

**Issue by frameproject** [ageron/handson-ml#205]
Bonjour Aurélien,

I have a question about linear regression (02_end_to_end_machine_learning_project.ipynb)
In chapter 2.6.1 (in the book) the linear regression is calculated from the housing_prepared and housing_label variables. A linear regression is of the form y = teta0 + teta1 X1 + teta2 X2 ... tetan Xn.
1) I do not understand what represents the variables X1, X2 ... latititude, longitude ...? and teta the parameters of the function?
2) What is the average price of housing in this case?
3) How to display the teta parameters from the regression (lin_reg.coef_? This gives a 4x4 matrix)

Thank you for your answer, I know you are very busy.

**Comment by frameproject** [comment]
Aurélien, 

I may have a solution:

from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(housing_prepared, housing_labels)
print(lin_reg.coef_)
intercept = lin_reg.intercept_
print("The intercept for our model is {}".format(intercept))

[-55650.4116403  -56716.45236929  13732.83841856  -1933.1277138
   7330.04062103 -45708.26306673  45455.47519691  74714.39134154
   6605.12802802   1042.95709453   9249.75886697 -18016.52432168
 -55219.15208555 110357.78363967 -22479.84008184 -14642.2671506 ]

The intercept for our model is 236914.997328

But I do not know what the parameters represent.

Fred

**Comment by ageron** [comment]
Hi @frameproject ,

>>> I do not understand what represents the variables X1, X2 ... latititude, longitude ...? and teta the parameters of the function?

x1 is the first input feature of one instance, in this problem it's the district's longitude.
x2 is the second input feature (latitude).
theta0 is the bias (also called intercept): it's a value that gets added to every prediction, regardless of the input features.
theta1 is the first feature's weight (in this case, the weight of the longitude): this is a model parameter, it's what we are trying to find during training.  Usually, if the corresponding feature is important, this weight will be important (assuming the input features were scaled), and conversely, if the feature is not important, then the weight will be small.

>>> What is the average price of housing in this case?

The mean house value is what we are trying to predict. If we are using a linear model, our prediction is:

y_pred = theta0 + theta1 * x1 + theta2 * x2 + ...

We actually have the correct value `y` for all the instances in the dataset, so we can compute the prediction error:

error = y_pred - y

The squared error is of course:

squared_error = error ** 2

We could compute this squared error over all the instances in the training set, or all the instances in the test set, and compute the mean: this would give us the mean squared error.  The goal of training is to minimize the MSE on the training set, in the hope that it will also minimize the MSE on new instances that are not in the training set.  Once we have found the optimal thetas, we can measure the generalization error by computing the MSE on the test set. This tells us how well our model can generalize to instances it never saw before.

>>> How to display the teta parameters from the regression (lin_reg.coef_? This gives a 4x4 matrix)

`lin_reg.coef_` should contain all the feature weights (theta1, theta2, ...).  The bias (theta0) is stored separately: `lin_reg.intercept_`.  So `lin_reg.coef_` should only contain one value per input feature. If there are 16 input features, then there should be 16 values in `lin_reg.coef_`, but it should be a 16x1 matrix (that's a column vector), not a 4x4 matrix.

Hope this helps,
Aurélien

**Comment by frameproject** [comment]
Many thanks, merci beaucoup !
